src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -340,565 +342,197 @@ void EspIdfWiFiBase::mdns_service_remove(const char *service)
 
-        if (results)
+#if !defined(ESP_IDF_PREDEF_NETIF)
+    {
+        // Check if we need to (re-)enable advertising and disable the STA
+        // interface.
+        OSMutexLock locker(&lock_);
+        if (mdnsAdvInhibited_)
         {
-            // Only return one result.
-            struct addrinfo *addr_info =
-                (struct addrinfo*)memp_malloc(MEMP_NETDB);
-            if (addr_info == nullptr)
-            {
-                return EAI_MEMORY;
-            }
-            memset(addr_info, 0, total_size);
-                    addr_info->ai_addrlen = sizeof(struct sockaddr_storage);
-            addr_info->ai_addr = (struct sockaddr*)((addr_info) + 1);
-            struct sockaddr_in *addr_in =
-                (struct sockaddr_in*)addr_info->ai_addr;
-#if defined(ESP_IDF_WIFI_IPV6)
-            struct sockaddr_in6 *addr_in6 =
-                (struct sockaddr_in6*)addr_info->ai_addr;
-#endif
-            mdns_ip_addr_t *current = results->addr;
-            while (current)
-            {
-                if (current->addr.type == ESP_IPADDR_TYPE_V4)
-                {
-                    addr_info->ai_family = AF_INET;
-                    addr_in->sin_family = AF_INET;
-                    addr_in->sin_port = htons(results->port);
-                    addr_in->sin_len = sizeof(struct sockaddr_in);
-                    addr_in->sin_addr.s_addr = current->addr.u_addr.ip4.addr;
-                    *addr = addr_info;
-                    break;
-                }
-#if defined(ESP_IDF_WIFI_IPV6)
-                else if (current->addr.type == ESP_IPADDR_TYPE_V6)
-                {
-                    addr_info->ai_family = AF_INET6;
-                    addr_in6->sin6_family = AF_INET6;
-                    addr_in6->sin6_port = htons(results->port);
-                    addr_in6->sin6_len = sizeof(struct sockaddr_in6);
-                    memcpy(addr_in6->sin6_addr.un.u32_addr,
-                        current->addr.u_addr.ip6.addr,
-                        sizeof(addr_in6->sin6_addr.un.u32_addr));
-                    *addr = addr_info;
-                    break;
-                }
-#else
-#endif
-                current = current->next;
-            }
-            if (*addr == nullptr)
-            {
-                // No result actually found.
-                freeaddrinfo(addr_info);
-            }
-            else
+            // We do not need to check for valid staIface_ or the
+            // mdnsAdvInhibitSta_ flag since we can only get here if it both
+            // were previously valid/set before starting the query.
+
+            // Update the timestamp for the last scan.
+            mdnsLookupTimestamp_ = OSTime::get_monotonic();
+            // Disable mDNS on STA. Note that the STA interface will also be
+            // unregistered, equivalent of calling mdns_unregister(staIface_).
+            ESP_ERROR_CHECK(
+                mdns_netif_action(staIface_, MDNS_EVENT_DISABLE_IP4));
```

Comment:
This is potentially setting us up for a race condition, though it is unlikely to actually occur. The netif actions and mdns remove/add are guaranteed to be sequential because they get queued up, in order, for the same worker thread to execute on. However, the `mdns_netif_action()` makes a synchronous check that the interface is registered before continuing. This could setup the following sequence:

1. `mdns_netif_action(staIface_, MDNS_EVENT_DISABLE_IP4)` is queue up
2. ...return from `mdns_lookup()`
3. ...call `mdsn_look()` again...
4. `mdns_register_netif(staIface_); ESP_ERROR_CHECK(mdns_netif_action(staIface_, MDNS_EVENT_ENABLE_IP4));` is called before the mDNS worker thread can address the queue.